docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,517 @@
+the operation is completed and the annotation to be removed.
+
+To achieve the restriction only gardenlet to create or refresh tokens in the
+seed cluster, but not any other controller running there, a new CRD
+`WorkloadIdentityBinding` in the seed will be implemented. This CRD will refer
+to the `WorkloadIdentity` in the Garden cluster that is source of the token, as
+well as specify target store where the tokens to be written. Gardenlet will be
+the only component having write access to this CRD, while extension controllers
+and other components relying on workload identities should request no more than
+read access. This `WorkloadIdentityBinding` CRD will be reconciled by a
+dedicated controller named `workloadidentity-refresher` from Gardenlet
+responsible to request OIDC tokens and write them into the target store. The
+workload identity `spec.config` will be also written into the target store.
+
+At the time this GEP is written, it is envisioned only secrets to be used as
+store targets. The token will be available on the data key
+`workloadIdentityToken` in the secret, while the service provider config will be
+at `workloadIdentityConfig`. These secrets will be labeled with
+`workloadidentity.authentication.gardener.cloud/provider=(aws|gcp|...)` so that
+the extensions can easily select them and make adjustments via admission
+webhooks, e.g. transform the service provider config and the token into
+canonical form usable by the respective service provider SDK.
+
+The secret `cloudprovider` will be reused as target store when the shoot is
+using workload identity tokens as infrastructure credentials, the flow should be
+as follow:
+
+1. Shoot reconciler creates/updates a `WorkloadIdentityBinding` resource in the
+   seed based on the configuration of the `WorkloadIdentity` and the shoot spec.
+   The name of the `WorkloadIdentityBinding` resource will be `cloudidentity`
+   while the name of the target store secret `cloudprovider`.
+1. `workloadidentity-refresher` receives event for `WorkloadIdentityBinding`
+   resource. It parses the token from the target store and if the token is still
+   valid, it just applies the provider config to the target store, otherwise a
+   new token is requested from the Gardener API and written to the store
+   together with the provider config.
+1. The status of the `WorkloadIdentityBinding` is updated with the "issued at"
+   and "expires at" timestamps of the token.
+1. The `WorkloadIdentityBinding` is requeued for reconciliation again when the
+   token will be suitable for renewal.
+
+A sample `WorkloadIdentityBinding.extensions.gardener.cloud` resource will look like:
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: WorkloadIdentityBinding
+metadata:
+  name: cloudidentity
+  namespace: shoot--local--foo
+spec:
+  type: aws
+  config:
+    apiVersion: aws.authentication.gardener.cloud/v1alpha1
+    kind: Config
+    iamRoleARN: arn:aws:iam::112233445566:role/gardener-dev
```

Comment:
`gardener-extension-validator-(aws|gcp|alicloud|azure|...)` will admit this config. They will parse and validate the inlined `Config` resource. Each extension will implement own semantic validation of the resource as it differs for each extension.

Reply:
I assume the the extensions need to do something else to weave in the information from the `Config`. Could you elaborate on how this implementation is planned in the text?

I think it would be helpful to have some details about the entire procedure in the text and ideally a sequence diagram (streamlined to the essential steps) to understand the flow e2e, from `WorkloadIdentity` creation to credentials usage by the provider extension during shoot reconcile.

Reply:
I will enhance the flow description to cover the end-to-end usage of the feature, it will consider how the token will be used by actual component, e.g. cloud-controller-manager.